airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java

```
@@ -22,97 +24,95 @@
-  Class<?> getContainerClass();
+  private record ContainerKey(Class<? extends ContainerFactory> clazz, DockerImageName imageName, List<String> methods) {};
 
-  /**
-   * Returns a shared instance of the testcontainer.
-   */
-  default C shared(String imageName, String... methods) {
-    final String mapKey = Stream.concat(
-        Stream.of(imageName, this.getClass().getCanonicalName()),
-        Stream.of(methods))
-        .collect(Collectors.joining("+"));
-    return Singleton.getOrCreate(mapKey, this);
-  }
+  private static class ContainerOrException {
 
-  /**
-   * This class is exclusively used by {@link #shared(String, String...)}. It wraps a specific shared
-   * testcontainer instance, which is created exactly once.
-   */
-  class Singleton<C extends JdbcDatabaseContainer<?>> {
+    private final Supplier<GenericContainer<?>> containerSupplier;
+    private volatile RuntimeException exception = null;
+    private volatile GenericContainer<?> _container = null;
 
-    static private final Logger LOGGER = LoggerFactory.getLogger(Singleton.class);
-    static private final ConcurrentHashMap<String, Singleton<?>> LAZY = new ConcurrentHashMap<>();
+    ContainerOrException(Supplier<GenericContainer<?>> containerSupplier) {
+      this.containerSupplier = containerSupplier;
+    }
 
-    @SuppressWarnings("unchecked")
-    static private <C extends JdbcDatabaseContainer<?>> C getOrCreate(String mapKey, ContainerFactory<C> factory) {
-      final Singleton<?> singleton = LAZY.computeIfAbsent(mapKey, Singleton<C>::new);
-      return ((Singleton<C>) singleton).getOrCreate(factory);
+    GenericContainer<?> container() {
+      if (exception == null && _container == null) {
+        synchronized (this) {
+          if (_container == null && exception == null) {
+            try {
+              _container = containerSupplier.get();
+            } catch (RuntimeException e) {
+              exception = e;
+            }
+          }
+        }
+      }
```

You still need to throw the exception here with `if (exception != null) { throw exception }`